opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,595 @@
+
+  remove_on_set_parameters_callback(dyn_params_handler_.get());
+  dyn_params_handler_.reset();
+  tf2_listener_.reset();
+
+  // Destroy bond connection
+  destroyBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+FollowingServer::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up %s", get_name());
+  tf2_buffer_.reset();
+  following_action_server_.reset();
+  controller_.reset();
+  vel_publisher_.reset();
+  filtered_dynamic_pose_pub_.reset();
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+FollowingServer::on_shutdown(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down %s", get_name());
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+template<typename ActionT>
+void FollowingServer::getPreemptedGoalIfRequested(
+  typename std::shared_ptr<const typename ActionT::Goal> goal,
+  const std::unique_ptr<nav2_util::SimpleActionServer<ActionT>> & action_server)
+{
+  if (action_server->is_preempt_requested()) {
+    goal = action_server->accept_pending_goal();
+  }
+}
+
+template<typename ActionT>
+bool FollowingServer::checkAndWarnIfCancelled(
+  std::unique_ptr<nav2_util::SimpleActionServer<ActionT>> & action_server,
+  const std::string & name)
+{
+  if (action_server->is_cancel_requested()) {
+    RCLCPP_WARN(get_logger(), "Goal was cancelled. Cancelling %s action", name.c_str());
+    return true;
+  }
+  return false;
+}
+
+template<typename ActionT>
+bool FollowingServer::checkAndWarnIfPreempted(
+  std::unique_ptr<nav2_util::SimpleActionServer<ActionT>> & action_server,
+  const std::string & name)
+{
+  if (action_server->is_preempt_requested()) {
+    RCLCPP_WARN(get_logger(), "Goal was preempted. Cancelling %s action", name.c_str());
+    return true;
+  }
+  return false;
+}
+
+void FollowingServer::followObject()
+{
+  std::lock_guard<std::mutex> lock(*dynamic_params_lock_);
+  action_start_time_ = this->now();
+  rclcpp::Rate loop_rate(controller_frequency_);
+
+  auto goal = following_action_server_->get_current_goal();
+  auto result = std::make_shared<FollowObject::Result>();
+
+  if (!following_action_server_ || !following_action_server_->is_server_active()) {
+    RCLCPP_DEBUG(get_logger(), "Action server unavailable or inactive. Stopping.");
+    return;
+  }
+
+  if (checkAndWarnIfCancelled(following_action_server_, "follow_object")) {
+    following_action_server_->terminate_all();
+    return;
+  }
+
+  getPreemptedGoalIfRequested(goal, following_action_server_);
+  num_retries_ = 0;
+
+  try {
+    RCLCPP_INFO(
+      get_logger(),
+      "Attempting to follow object at position (%0.2f, %0.2f).",
+      goal->object_pose.pose.position.x, goal->object_pose.pose.position.y);
+
+    // Construct initial estimate of where the object is located in fixed_frame
+    auto object_pose = goal->object_pose;
+    object_pose.header.stamp = rclcpp::Time(0);
```

Comment:
Use an actual rclcpp::Time from `now()` and add a transform tolerance of like 0.1 (parameterize it)